### PR TITLE
[JIT-460] When claiming mev tips, skip accounts that won't have min rent exempt amount after claiming

### PR DESCRIPTION
**Pull request by esemeniuc**
#### Problem
See ticket: https://linear.app/jito/issue/JIT-460/when-claiming-mev-tips-skip-accounts-that-wont-have-min-rent-exempt

#### Summary of Changes
- Reduce TDA rpc calls
- Flatten error handling
- Add balance and rent checks
- Add logging for debugging transient errors

### CODE REVIEW

File: tip-distributor/src/claim_mev_workflow.rs [hunk 1]
```diff
@@ -55,13 +55,26 @@ pub fn claim_mev_tips(
...
-            .expect("read blockhash");
-
+        let blockhash = rpc_client.get_latest_blockhash().await.expect("read blockhash");
+        let balance = rpc_client.get_balance(&keypair.pubkey()).await.expect("failed to get balance");
+
+        // heuristic to make sure we have enough funds to cover the rent costs if jito has many validators
+        {
+            let node_count = merkle_trees.generated_merkle_trees.iter().flat_map(|tree| &tree.tree_nodes).count();
+            let min_rent = rpc_client.get_minimum_balance_for_rent_exemption(ClaimStatus::SIZE).await.expect("Failed to calculate min rent");
+            assert!(balance >= node_count as u64 * min_rent);
```
Comment (buffalu) [inline tip-distributor/src/claim_mev_workflow.rs:65]: we could add in a fee estimation too. like merkle_trees.generated_merkle_trees * tree.tree_nodes * FEE_PER_SIGNATURE

Comment (buffalu) [reply tip-distributor/src/claim_mev_workflow.rs:65]: this seems perfect to me though, will overestimate but seems fine to alert earlier than later

File: tip-distributor/src/claim_mev_workflow.rs [hunk 2]
```diff
@@ -70,71 +83,61 @@ pub fn claim_mev_tips(
...
+
+                let claim_status_acc_result = rpc_client.get_account(&claim_status_pubkey).await;
+                if claim_status_acc_result.is_ok() {
+                    debug!("claim status account already exists: {:?}", claim_status_pubkey);
+                    continue;
                 }
+                if !(matches!(claim_status_acc_result.unwrap_err().kind(), ErrorKind::RpcError(RpcError::ForUser(_)))) {
+                    panic!("Invalid RPC Error");
+                }
+                info!("claiming for public key: {:?}", node.claimant);
```
Comment (buffalu) [inline tip-distributor/src/claim_mev_workflow.rs:104]: nit: move this below the min_rent check

File: tip-distributor/src/lib.rs [hunk 3]
```diff
@@ -430,10 +430,10 @@ pub async fn send_transactions_with_retry(
         }
     }
 
-    assert!(
-        transactions_to_send.is_empty(),
-        "all transactions failed to send"
-    );
+    if !transactions_to_send.is_empty() {
+        error!("still have {} txns to send", transactions_to_send.len());
+        panic!("failed to send all transactions");
```
Comment (buffalu) [inline tip-distributor/src/lib.rs:435]: can you combine these?

**Review by buffalu — APPROVED** [review]
few tiny nits but lgtm 🚢 🚢 🚢

File: tip-distributor/src/claim_mev_workflow.rs [hunk 4]
```diff
@@ -55,13 +55,25 @@ pub fn claim_mev_tips(
     let mut transactions = Vec::new();
 
     runtime.block_on(async move {
-        let blockhash = rpc_client
-            .get_latest_blockhash()
-            .await
-            .expect("read blockhash");
-
+        let blockhash = rpc_client.get_latest_blockhash().await.expect("read blockhash");
```
Comment (segfaultdoc) [inline tip-distributor/src/claim_mev_workflow.rs:58]: curious if reformat caused by your specific IDE settings

File: tip-distributor/src/claim_mev_workflow.rs [hunk 5]
```diff
@@ -70,71 +82,60 @@ pub fn claim_mev_tips(
...
+                    continue;
+                }
+
+                let claim_status_acc_result = rpc_client.get_account(&claim_status_pubkey).await;
+                if claim_status_acc_result.is_ok() {
+                    debug!("claim status account already exists: {:?}", claim_status_pubkey);
+                    continue;
                 }
+                if !(matches!(claim_status_acc_result.unwrap_err().kind(), ErrorKind::RpcError(RpcError::ForUser(_)))) {
+                    panic!("Invalid RPC Error");
```
Comment (segfaultdoc) [inline tip-distributor/src/claim_mev_workflow.rs:101]: wdyt about matching on the error and printing it here?

File: tip-distributor/src/claim_mev_workflow.rs (hunk 4 above)
Comment (esemeniuc) [reply tip-distributor/src/claim_mev_workflow.rs:58]: not sure tbh, ran `cargo +nightly fmt` afterwards

File: tip-distributor/src/claim_mev_workflow.rs (hunk 5 above)
Comment (buffalu) [reply tip-distributor/src/claim_mev_workflow.rs:101]: can log claim_status_acc_result here

File: tip-distributor/src/claim_mev_workflow.rs [hunk 6]
```diff
@@ -70,71 +82,60 @@ pub fn claim_mev_tips(
...
+                }
+                let account = rpc_client.get_account(&node.claimant).await.expect("Failed to fetch account");
+                let min_rent = rpc_client.get_minimum_balance_for_rent_exemption(account.data.len()).await.expect("Failed to calculate min rent");
+                if node.amount < min_rent {
```
Comment (buffalu) [inline tip-distributor/src/claim_mev_workflow.rs:105]: from slack discussion: needs to be node.amount + account.balance < min_rent

File: tip-distributor/src/claim_mev_workflow.rs [hunk 7]
```diff
@@ -70,71 +82,60 @@ pub fn claim_mev_tips(
...
                     );
+                    continue;
+                }
+
+                let claim_status_acc_result = rpc_client.get_account(&claim_status_pubkey).await;
+                if claim_status_acc_result.is_ok() {
+                    debug!("claim status account already exists: {:?}", claim_status_pubkey);
+                    continue;
                 }
+                if !(matches!(claim_status_acc_result.unwrap_err().kind(), ErrorKind::RpcError(RpcError::ForUser(_)))) {
```
[the remark this answers is not in the document]
Comment (esemeniuc) [reply tip-distributor/src/claim_mev_workflow.rs]: updated

File: tip-distributor/src/claim_mev_workflow.rs (hunk 5 above)
Comment (esemeniuc) [reply tip-distributor/src/claim_mev_workflow.rs]: logged now

File: tip-distributor/src/claim_mev_workflow.rs (hunk 6 above)
Comment (esemeniuc) [reply tip-distributor/src/claim_mev_workflow.rs]: updated

**Pull request merged**